### PR TITLE
Handle generated resources from plugins

### PR DESCRIPTION
**Pull request by neonichu**
rdar://89693335

### CODE REVIEW

**Comment by neonichu** [comment]
Biggest question I have: since we don't support generated sources for Clang targets, yet, I am assuming we do the same for resources for now and handle support for Clang targets separately?

File: Sources/Build/BuildPlan.swift
Comment (tomerd) [inline Sources/Build/BuildPlan.swift:693]: observability?

File: Sources/Build/BuildPlan.swift
Comment (tomerd) [inline Sources/Build/BuildPlan.swift:706]: should this throw / warn?

File: Sources/Build/BuildPlan.swift
Comment (tomerd) [inline Sources/Build/BuildPlan.swift:709]: should this throw / warn?

File: Sources/Build/BuildPlan.swift
Comment (tomerd) [inline Sources/Build/BuildPlan.swift:700]: should this throw / warn?

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (tomerd) [inline Sources/PackageLoading/TargetSourcesBuilder.swift:194]: observabilityScope :(

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (tomerd) [inline Sources/PackageLoading/TargetSourcesBuilder.swift:293]: need to be addressed?

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (tomerd) [inline Sources/PackageLoading/TargetSourcesBuilder.swift:281]: need to be addressed?

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (neonichu) [reply Sources/PackageLoading/TargetSourcesBuilder.swift:293]: Yep, was just trying to get something roughly working to see if this approach makes sense to everyone.

**Comment by neonichu** [comment]
Oh, there's one major limitation right now: because we do not retain information from the manifest about non-existent files during the initial pass of `TargetSourcesBuilder`, this will only work for built-in resource types. 

Maybe we should treat any unknown resources as to be processed right now? The main reason we don't include anything but certain known types automatically was that we wanted to prevent people from accidentally shipping files with their products, but since any files here are basically already opt-in through depending on a given plugin, that might not be a concern?

File: Sources/Build/BuildPlan.swift
Comment (abertelrud) [inline Sources/Build/BuildPlan.swift:565]: How come asking for the target's bundle doesn't work here?

File: Sources/Build/BuildPlan.swift
Comment (abertelrud) [reply Sources/Build/BuildPlan.swift:709]: Probably warn, I'd say.  That's what we do for other unhandled files.

File: Sources/Build/BuildPlan.swift
Comment (abertelrud) [inline Sources/Build/BuildPlan.swift:722]: 👍

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (abertelrud) [reply Sources/PackageLoading/TargetSourcesBuilder.swift:281]: Yes, I think this is just a draft at this point to get early feedback.  Agree it needs to be addressed before this is merged.

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (abertelrud) [reply Sources/PackageLoading/TargetSourcesBuilder.swift:293]: I think it does.  This is along the lines of what I was thinking, though I hadn't looked into it as closely yet as you have.  Eventually we probably want to rework TargetSourcesBuilder to carry through some of the manifest information, as you said in one of the comments, but I think there is value in some support here even if all cases aren't yet handled.

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (abertelrud) [inline Sources/PackageLoading/TargetSourcesBuilder.swift:192]: Is this the one that clients of libSwiftPM would now call for output files from plugins?

File: Sources/Build/BuildPlan.swift
Comment (neonichu) [reply Sources/Build/BuildPlan.swift:565]: Currently, the bundle name is set to `nil` when there are no resources during the initial package construction, I'll be changing that.

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (neonichu) [reply Sources/PackageLoading/TargetSourcesBuilder.swift:192]: I think so

File: Sources/Build/BuildPlan.swift
Comment (abertelrud) [reply Sources/Build/BuildPlan.swift:565]: Ah, cool.  That sounds like a good approach.

**Comment by abertelrud** [comment]
> Biggest question I have: since we don't support generated sources for Clang targets, yet, I am assuming we do the same for resources for now and handle support for Clang targets separately?

Yes, I think that makes sense.  It's unfortunate that there's such a split between Clang targets and Swift targets in SwiftPM, causing overlap for the things that aren't difference.  But I think it's a good idea to tackle this piecemeal, as you're doing.

File: Sources/Build/BuildPlan.swift
Comment (neonichu) [reply Sources/Build/BuildPlan.swift]: I changed it to be ignored since it looks like `TargetSourcesBuilder` itself would already warn in this case.

File: Sources/Build/BuildPlan.swift
Comment (neonichu) [reply Sources/Build/BuildPlan.swift]: Don't have a strong opinion, but I think error seems appropriate since a plugin is essentially broken/incomplete if it generates files that aren't handled at all. I can see the argument that it should be a warning because the client of a plugin should maybe be able to use a broken plugin anyway?

File: Sources/Build/BuildPlan.swift
Comment (neonichu) [reply Sources/Build/BuildPlan.swift]: Don't have a strong opinion, but I think error seems appropriate since a plugin is essentially broken/incomplete if it generates files that aren't handled at all. I can see the argument that it should be a warning because the client of a plugin should maybe be able to use a broken plugin anyway?

**Comment by neonichu** [comment]
OK, I think this is basically ready for review now, just need to add some tests.

**Comment by neonichu** [comment]
@swift-ci please smoke test

File: Sources/Build/BuildPlan.swift
Comment (abertelrud) [reply Sources/Build/BuildPlan.swift]: Yeah, that's a good point, it's different from just having a file that I added to my sources but I don't want to be blocked from building because of.  I don't feel strongly about this though.  I think it would be fine to start with a warning.

File: Sources/Build/BuildPlan.swift
Comment (abertelrud) [reply Sources/Build/BuildPlan.swift]: Same comment as above.  I don't feel strongly.

File: Sources/Build/BuildPlan.swift
Comment (abertelrud) [inline Sources/Build/BuildPlan.swift:723]: This is good as a start — a future PR could probably consolidate these into a single error with an associated list of headers, but can definitely be a future PR.

**Review by abertelrud — APPROVED** [review]
LGTM (modulo the need for tests as you mentioned).  Thank you for taking this on!

File: Sources/Build/BuildPlan.swift
Comment (tomerd) [inline Sources/Build/BuildPlan.swift:184]: wonder if we can expose that at the target level so we dont have to drill down so deep?

File: Sources/Build/BuildPlan.swift
Comment (tomerd) [inline Sources/Build/BuildPlan.swift:564]: nit: in this cases I like to use `.none` but its a matter of personal style

File: Sources/Build/BuildPlan.swift
Comment (tomerd) [inline Sources/Build/BuildPlan.swift:673]: \o/

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (tomerd) [inline Sources/PackageLoading/TargetSourcesBuilder.swift:198]: nit: format func arguments into new lines

File: Tests/PackageGraphTests/TargetTests.swift
Comment (tomerd) [inline Tests/PackageGraphTests/TargetTests.swift:23]: nit: new line

File: Sources/Build/BuildPlan.swift
Comment (neonichu) [reply Sources/Build/BuildPlan.swift]: Switching to a warning for now.

File: Sources/Build/BuildPlan.swift
Comment (neonichu) [reply Sources/Build/BuildPlan.swift]: Switching to a warning for now.

File: Sources/Build/BuildPlan.swift
Comment (neonichu) [reply Sources/Build/BuildPlan.swift:184]: I think so, but I'll leave Clang target untouched for this PR, we can clean it up once we are adding support for Clang targets.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
```
/private/var/folders/pf/p5g2vb2j0yx1x6vlmxdppc700000gn/T/Miscellaneous_PluginGeneratedResources.qrxDtx/Miscellaneous_PluginGeneratedResources/.build/plugins/cache/Generator>, <output:
'dyld: lazy symbol binding failed: can'\''t resolve symbol _$ss13_runAsyncMainyyyyYaKcF in /private/var/folders/pf/p5g2vb2j0yx1x6vlmxdppc700000gn/T/Miscellaneous_PluginGeneratedResources.qrxDtx/Miscellaneous_PluginGeneratedResources/.build/plugins/cache/Generator because dependent dylib /usr/lib/swift/libswift_Concurrency.dylib could not be loaded
dyld: can'\''t resolve symbol _$ss13_runAsyncMainyyyyYaKcF in /private/var/folders/pf/p5g2vb2j0yx1x6vlmxdppc700000gn/T/Miscellaneous_PluginGeneratedResources.qrxDtx/Miscellaneous_PluginGeneratedResources/.build/plugins/cache/Generator because dependent dylib /usr/lib/swift/libswift_Concurrency.dylib could not be loaded
'>
```

**Comment by neonichu** [comment]
Self-hosted is running macOS 11.5.2, so we shouldn't be trying to find the concurrency libs in the OS, I think. Also odd that the smoke test passes just fine. Maybe an issue with the superior toolchain?

**Comment by neonichu** [comment]
It is using 

```
swift --version
swift-driver version: 1.26.7 Apple Swift version 5.5 (swiftlang-1300.0.27.6 clang-1300.0.27.2)
Target: x86_64-apple-macosx11.0
```

So maybe we'll just disable the test for < 5.6

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by abertelrud** [comment]
> It is using
> 
> ```
> swift --version
> swift-driver version: 1.26.7 Apple Swift version 5.5 (swiftlang-1300.0.27.6 clang-1300.0.27.2)
> Target: x86_64-apple-macosx11.0
> ```
> 
> So maybe we'll just disable the test for < 5.6

Oh I see!  Yes, this needs at least 5.6.

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (abertelrud) [inline Sources/PackageLoading/TargetSourcesBuilder.swift:503]: Nit: Maybe "Only Swift is supported for generated plugin source files at this time" to be similar to the other messages here?

File: Sources/PackageModel/Target.swift
Comment (abertelrud) [inline Sources/PackageModel/Target.swift:216]: Any compatibility issues or is this use of Codable just a cache?

File: Sources/Build/BuildPlan.swift
Comment (tomerd) [inline Sources/Build/BuildPlan.swift:575]: should we add `target.resources` pointing to `target.underlyingTarget.resources`?

File: Sources/Build/BuildPlan.swift
Comment (tomerd) [inline Sources/Build/BuildPlan.swift:561]: should we add `target.potentialBundleName` pointing to `target.underlyingTarget.potentialBundleName`?

File: Sources/Build/BuildPlan.swift
Comment (tomerd) [inline Sources/Build/BuildPlan.swift:253]: should we add `target.bundleName` pointing to `target.underlyingTarget.bundleName`?

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (tomerd) [inline Sources/PackageLoading/TargetSourcesBuilder.swift:128]: nit: `(toolsVersion <= .v5_4 `

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (tomerd) [inline Sources/PackageLoading/TargetSourcesBuilder.swift:198]: intentional that declaredResources is empty array and declaredSources is nil? should they be consistent?

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (tomerd) [inline Sources/PackageLoading/TargetSourcesBuilder.swift:486]: guard?

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (tomerd) [inline Sources/PackageLoading/TargetSourcesBuilder.swift:496]: we can also make .none a case in the switch below instead of explicitly handling it here

**Review by tomerd — COMMENTED** [review]
looks good, some style nits

File: Sources/Build/BuildPlan.swift
Comment (abertelrud) [reply Sources/Build/BuildPlan.swift:184]: That makes sense.  When we get to them though, I think it would make sense to have a wrapper as for Swift targets (should that perhaps be added already now, except that for Clang targets it just returns the non-generated ones?).

**Review by abertelrud — APPROVED** [review]

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (neonichu) [reply Sources/PackageLoading/TargetSourcesBuilder.swift:496]: We could, but I think it's easier to understand this way this way than doing any weird dance with `fallthrough` in the `switch` to avoid duplication.

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (neonichu) [reply Sources/PackageLoading/TargetSourcesBuilder.swift:486]: Feels wrong to use `guard` for a legitimate case?

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (neonichu) [reply Sources/PackageLoading/TargetSourcesBuilder.swift:198]: Yah, it's intentional since they get different treatment in the existing code.

File: Sources/PackageModel/Target.swift
Comment (neonichu) [reply Sources/PackageModel/Target.swift:216]: My understanding is that it is just a cache.

File: Sources/Build/BuildPlan.swift
Comment (neonichu) [reply Sources/Build/BuildPlan.swift:575]: Maybe, but reaching into `target.underlyingTarget` is a common pattern, so I think if we want to change that, that should be a separate PR.

File: Sources/Build/BuildPlan.swift
Comment (neonichu) [reply Sources/Build/BuildPlan.swift:253]: see above

File: Sources/Build/BuildPlan.swift
Comment (neonichu) [reply Sources/Build/BuildPlan.swift:561]: see above

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (tomerd) [reply Sources/PackageLoading/TargetSourcesBuilder.swift:496]: would it not just be `case .none, .copy, .processResource:`?

Comment (neonichu) [reply Sources/PackageLoading/TargetSourcesBuilder.swift:496]: But we have to swap the rule to the right one before doing the code for copy/process.

Comment (neonichu) [reply Sources/PackageLoading/TargetSourcesBuilder.swift:496]: I guess we could do something like

```
let actualRule = rule == .none ? .processResource(localization: .none) : rule
```

or pull the `if` statement into the `switch`, but I think that's more difficult to understand.

File: Sources/Build/BuildPlan.swift
Comment (tomerd) [reply Sources/Build/BuildPlan.swift:575]: probably worth a dedicated cleanup then. should we add a ticket to track?

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (tomerd) [reply Sources/PackageLoading/TargetSourcesBuilder.swift:496]: I missed the fact we need to send the rule into `Self.resource`. we could tease `.none` and `.processResource` case handling apart from `.copy`and pass in `.processResource` into `Self.resource` for both. that would feel the cleanest imo.

in any case its not a blocker for this PR

File: Sources/Build/BuildPlan.swift
Comment (neonichu) [reply Sources/Build/BuildPlan.swift:575]: rdar://92180880 (Clean up use of `target.underlyingTarget`)

File: Sources/PackageLoading/TargetSourcesBuilder.swift
Comment (neonichu) [reply Sources/PackageLoading/TargetSourcesBuilder.swift:496]: I think I prefer the current structure since it calls more attention to the `.none` handling being a special case before we get to the regular handling of file types, instead of merging the two.

**Comment by neonichu** [comment]
@swift-ci please smoke test

**Comment by neonichu** [comment]
@swift-ci please smoke test linux

**Pull request merged**
